cmd/snap-confine/udev-support.c

```diff
@@ -18,290 +18,368 @@
+	}
+}
 
-	if (udev_enumerate_add_match_tag(udev_s->devices, udev_s->tagname) < 0)
-		die("udev_enumerate_add_match_tag");
+/**
+ * Allow access to /dev/uhid.
+ *
+ * Currently /dev/uhid isn't represented in sysfs, so add it to the device
+ * cgroup if it exists and let AppArmor handle the mediation.
+ **/
+static void sc_udev_allow_uhid(int devices_allow_fd)
+{
+	struct stat sbuf;
 
-	if (udev_enumerate_scan_devices(udev_s->devices) < 0)
-		die("udev_enumerate_scan failed");
+	if (stat("/dev/uhid", &sbuf) == 0) {
+		sc_dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			   minor(sbuf.st_rdev));
+	}
+}
 
-	udev_s->assigned = udev_enumerate_get_list_entry(udev_s->devices);
-	if (udev_s->assigned == NULL)
-		rc = -1;
+/**
+ * Allow access to assigned devices.
+ *
+ * The snapd udev security backend uses udev rules to tag matching devices with
+ * tags corresponding to snap applications. Here we interrogate udev and allow
+ * access to all assigned devices.
+ **/
+static void sc_udev_allow_assigned(int devices_allow_fd, struct udev *udev,
+				   struct udev_list_entry *assigned)
+{
+	for (struct udev_list_entry * entry = assigned; entry != NULL;
+	     entry = udev_list_entry_get_next(entry)) {
+		const char *path = udev_list_entry_get_name(entry);
+		if (path == NULL) {
+			die("udev_list_entry_get_name failed");
+		}
+		struct udev_device *device =
+		    udev_device_new_from_syspath(udev, path);
+		if (device == NULL) {
+			die("cannot find device from syspath %s", path);
+		}
+		dev_t devnum = udev_device_get_devnum(device);
+		char type_c = strstr(path, "/block/") != NULL ? 'b' : 'c';
+		sc_dprintf(devices_allow_fd, "%c %u:%u rwm", type_c,
+			   major(devnum), minor(devnum));
 
-	return rc;
+		udev_device_unref(device);
+	}
 }
 
-void snappy_udev_cleanup(struct snappy_udev *udev_s)
+static void sc_udev_setup_acls(int devices_allow_fd, int devices_deny_fd,
+			       struct udev *udev,
+			       struct udev_list_entry *assigned)
 {
-	// udev_s->assigned does not need to be unreferenced since it is a
-	// pointer into udev_s->devices
-	if (udev_s->devices != NULL)
-		udev_enumerate_unref(udev_s->devices);
-	if (udev_s->udev != NULL)
-		udev_unref(udev_s->udev);
+	/* Deny device access by default.
+	 *
+	 * Write 'a' to devices.deny to remove all existing devices that were added
+	 * in previous launcher invocations, then add the static and assigned
+	 * devices. This ensures that at application launch the cgroup only has
+	 * what is currently assigned. */
+	sc_dprintf(devices_deny_fd, "a");
+
+	/* Allow access to various devices. */
+	sc_udev_allow_common(devices_allow_fd);
+	sc_udev_allow_pty_slaves(devices_allow_fd);
+	sc_udev_allow_nvidia(devices_allow_fd);
+	sc_udev_allow_uhid(devices_allow_fd);
+	sc_udev_allow_assigned(devices_allow_fd, udev, assigned);
 }
 
-void setup_devices_cgroup(const char *security_tag, struct snappy_udev *udev_s)
+static char *sc_udev_mangle_security_tag(const char *security_tag)
```

Comment:
How about `sc_security_to_udev_tag`?

Reply:
Sure, do you mind adding a suggestion? I can look later today as well.

Reply:
Done